src/main/clojure/clara/rules/engine.cljc

```diff
@@ -713,16 +713,14 @@
       (mem/add-tokens! memory node join-bindings tokens)
       
       (cond
-        ;; If there are previously accumulated results to propagate, use them.  If
-        ;; this is the first time there are matching tokens, then the reduce will
-        ;; have to happen for the first time.  However, this reduce operation is
-        ;; independent of the specific tokens since the elements join to the tokens
-        ;; via pre-computed hash join bindings for this node.  So only reduce once,
-        ;; for all tokens.
+        ;; If there are previously accumulated results to propagate, use them.  If this is the
+        ;; first time there are matching tokens, then the reduce will have to happen for the
+        ;; first time.  However, this reduce operation is independent of the specific tokens
+        ;; since the elements join to the tokens via pre-computed hash join bindings for this
+        ;; node.  So only reduce once per group of facts, for all tokens.
```

Comment:
You might specifically note that we group on all bindings, not just the join bindings.